File: sdks/swift/core/Sources/VisionAgentsCore/SessionSocket.swift
import Foundation

/// The socket carrying one conversation.
///
/// This is hand-written because OpenAPI stops at the upgrade. It is an actor so that the read
/// loop and the sends cannot race over the task, and so that closing twice is harmless.
///
/// `URLSessionWebSocketTask` answers the router's pings itself, so there is no keepalive here.
/// What it does not do is notice a task that was torn down while the app was in the
/// background: that arrives as a read failure, which ends the event stream.
/// How big one frame may be. A transcript event carries a sentence, and a tool call carries the
/// arguments a model wrote, so a megabyte is generous.
private let maximumFrameSize = 1 << 20

public actor SessionSocket {
    private let url: URL
    private let headers: [String: String]
    private let urlSession: URLSession
    private var task: URLSessionWebSocketTask?
    private var reader: Task<Void, Never>?

    public init(url: URL, headers: [String: String], urlSession: URLSession = .shared) {
        self.url = url
        self.headers = headers
        self.urlSession = urlSession
    }

    /// Opens the socket and starts reading.
    ///
    /// The stream is returned rather than exposed as a property so that it cannot be attached
    /// to twice, and so there is no window between opening and attaching in which an event
    /// could be dropped: reading does not begin until the stream exists.
    public func open() -> AsyncThrowingStream<AgentEvent, any Error> {
        var request = URLRequest(url: url)
        for (name, value) in headers {
            request.setValue(value, forHTTPHeaderField: name)
        }

        let task = urlSession.webSocketTask(with: request)
        // A conversation's frames are a few kilobytes at most. Without a ceiling, a socket
        // that has gone wrong at the other end can make this process buffer without bound.
        task.maximumMessageSize = maximumFrameSize
        self.task = task
        task.resume()

        // Unbounded because dropping an event would lose a tool call or the end of a reply.
        // The producer is a network socket, so the queue is bounded in practice by what the
        // router sends.
        let (stream, continuation) = AsyncThrowingStream.makeStream(
            of: AgentEvent.self, bufferingPolicy: .unbounded)

        reader = Task { [weak self] in
            await self?.read(task: task, into: continuation)
        }

        return stream
    }

    /// Sends one command. Fails if the socket is not open.
    public func send(_ command: Command) async throws {
        guard let task else {
            throw AgentsError.socketClosed(code: 0, reason: "the socket is not open")
        }
        let encoded = try JSONEncoder().encode(command)
        do {
            try await task.send(.data(encoded))
        } catch {
            throw AgentsError.transport(error)
        }
    }

    /// Closes the socket. Safe to call more than once.
    public func close() {
        reader?.cancel()
        reader = nil
        task?.cancel(with: .normalClosure, reason: nil)
        task = nil
    }

    private func read(
        task: URLSessionWebSocketTask,
        into continuation: AsyncThrowingStream<AgentEvent, any Error>.Continuation
    ) async {
        let decoder = JSONDecoder()
        while !Task.isCancelled {
            let message: URLSessionWebSocketTask.Message
            do {
                message = try await task.receive()
            } catch {
                // A cancelled read is a close we asked for, not a failure to report.
                if Task.isCancelled {
                    continuation.finish()
                } else {
                    continuation.finish(
                        throwing: AgentsError.socketClosed(
                            code: task.closeCode.rawValue,
                            reason: (error as NSError).localizedDescription))
                }
                return
            }

            let payload: Data
            switch message {
            case .data(let data):
                payload = data
            case .string(let text):
                payload = Data(text.utf8)
            @unknown default:
                continue
            }

            // A frame this SDK cannot read is skipped rather than ending the conversation,
            // which is what the router does with a command it cannot read.
            if let event = try? decoder.decode(AgentEvent.self, from: payload) {
                continuation.yield(event)
            }
        }
        continuation.finish()
    }
}

File: sdks/kotlin/core/src/main/kotlin/io/getstream/visionagents/core/SessionSocket.kt
package io.getstream.visionagents.core

import io.ktor.client.plugins.websocket.DefaultClientWebSocketSession
import io.ktor.client.plugins.websocket.webSocketSession
import io.ktor.client.request.header
import io.ktor.websocket.CloseReason
import io.ktor.websocket.Frame
import io.ktor.websocket.close
import io.ktor.websocket.readBytes
import io.ktor.websocket.readText
import java.io.IOException
import java.util.concurrent.atomic.AtomicBoolean
import kotlinx.coroutines.channels.ClosedSendChannelException
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow

/**
 * The socket carrying one conversation.
 *
 * Hand-written because OpenAPI stops at the upgrade. OkHttp answers the router's pings itself,
 * so there is no keepalive here, and no JSON `"ping"` either.
 *
 * **No automatic reconnection, on purpose.** `respond` and `tool_result` are not idempotent,
 * and the protocol has no sequence number to resume from, so replaying after a reconnect would
 * duplicate turns and tool results. A socket that drops ends the flow with
 * [AgentsException.SocketClosed] and the caller decides.
 */
public class SessionSocket internal constructor(
    private val backend: Backend,
    private val sessionId: String,
    private val query: List<Pair<String, String>>,
) {
    private val opened = AtomicBoolean(false)
    private val closing = AtomicBoolean(false)
    @Volatile private var connection: DefaultClientWebSocketSession? = null

    /**
     * Connects, and returns the session's events.
     *
     * Returned rather than exposed as a property so that it cannot be read twice: one reader
     * per connection, because two readers take half the frames each. Nothing is missed between
     * connecting and collecting, since the router starts watching before the upgrade and the
     * frames wait in the connection until they are read.
     *
     * The flow completes when the session ends or [close] is called, and throws
     * [AgentsException.SocketClosed] when the socket goes away for any other reason.
     */
    public suspend fun open(): Flow<AgentEvent> {
        if (!opened.compareAndSet(false, true)) {
            throw AgentsException.Configuration("a session socket is opened once")
        }
        val headers = backend.headers()
        val url = backend.socketUrl("v1", "agents", "sessions", sessionId, "events", query = query)
        val ws = try {
            backend.http.webSocketSession(url) {
                headers.forEach { (name, value) -> header(name, value) }
            }
        } catch (e: IOException) {
            throw AgentsException.Transport(e)
        }
        connection = ws

        return flow {
            try {
                for (frame in ws.incoming) {
                    val text = when (frame) {
                        is Frame.Text -> frame.readText()
                        is Frame.Binary -> frame.readBytes().decodeToString()
                        else -> continue
                    }
                    AgentEvent.decode(text)?.let { emit(it) }
                }
            } catch (e: IOException) {
                if (closing.get()) return@flow
                // 1006 is what RFC 6455 reserves for a connection that ended with no close frame.
                throw AgentsException.SocketClosed(1006, e.message.orEmpty())
            }
            val reason = ws.closeReason.await()
            if (!closing.get() && reason != null && reason.code != CloseReason.Codes.NORMAL.code) {
                throw AgentsException.SocketClosed(reason.code.toInt(), reason.message)
            }
        }
    }

    /** Sends one command. Fails if the socket is not open. */
    public suspend fun send(command: Command) {
        val ws = connection ?: throw AgentsException.SocketClosed(0, "the socket is not open")
        try {
            ws.send(Frame.Text(command.encode()))
        } catch (_: ClosedSendChannelException) {
            throw AgentsException.SocketClosed(0, "the socket is closed")
        }
    }

    /** Closes the socket with a normal closure. Safe to call more than once. */
    public suspend fun close() {
        if (!closing.compareAndSet(false, true)) return
        connection?.close(CloseReason(CloseReason.Codes.NORMAL, ""))
    }
}
